UniversalAlgebra/Homomorphisms/Setoid.lagda

```diff
@@ -0,0 +1,307 @@
+  Goal 𝑓 a = (h ∘ g)((𝑓 ̂ 𝑨) a)     ≡⟨ ≡-cong h ( ghom 𝑓 a ) ⟩
+             h ((𝑓 ̂ 𝑩)(g ∘ a))     ≡⟨ hhom 𝑓 ( g ∘ a ) ⟩
+             (𝑓 ̂ 𝑪)(h ∘ g ∘ a)     ∎
+
+
+ ∘-is-hom : {f : A → B}{g : B → C}
+  →         is-homomorphism 𝑨 𝑩 f → is-homomorphism 𝑩 𝑪 g
+            -------------------------------------------------
+  →         is-homomorphism 𝑨 𝑪 (g ∘ f)
+
+ ∘-is-hom {f} {g} fhom ghom = ∥ ∘-hom (f , fhom) (g , ghom) ∥
+
+module _ {α ρ : Level} where
+
+ -- the identity homs
+ 𝒾𝒹 : (𝑨 : SetoidAlgebra α ρ) → hom 𝑨 𝑨
+ 𝒾𝒹 _ = id , λ 𝑓 a → refl
+
+ open Level
+
+ -- the lift hom
+ 𝓁𝒾𝒻𝓉 : {β : Level}(𝑨 : SetoidAlgebra α ρ) → hom 𝑨 (Lift-SetoidAlg 𝑨 β)
+ 𝓁𝒾𝒻𝓉 _ = lift , (λ 𝑓 a → refl)
+
+ -- the lower hom
+ 𝓁ℴ𝓌ℯ𝓇 : {β : Level}(𝑨 : SetoidAlgebra α ρ) → hom (Lift-SetoidAlg 𝑨 β) 𝑨
+ 𝓁ℴ𝓌ℯ𝓇 _ = (lower , λ 𝑓 a → refl)
+
+module _ {α ρᵃ : Level} {𝑨 : SetoidAlgebra α ρᵃ}
+         (ℓᵃ : Level)
+         {β ρᵇ : Level}{𝑩 : SetoidAlgebra β ρᵇ}
+         (ℓᵇ : Level)
+         where
+ open Level
+
+ Lift-hom : hom 𝑨 𝑩  →  hom (Lift-SetoidAlg 𝑨 ℓᵃ) (Lift-SetoidAlg 𝑩 ℓᵇ)
+
+ Lift-hom (f , fhom) = lift ∘ f ∘ lower , Goal
+  where
+  lA lB : SetoidAlgebra _ _
+  lA = Lift-SetoidAlg 𝑨 ℓᵃ
+  lB = Lift-SetoidAlg 𝑩 ℓᵇ
+
+  lABh : is-homomorphism lA 𝑩 (f ∘ lower)
+  lABh = ∘-is-hom lA {𝑩 = 𝑨}  𝑩 {lower}{f} (λ _ _ → refl) fhom
+
+  Goal : is-homomorphism lA lB (lift ∘ (f ∘ lower))
+  Goal = ∘-is-hom lA {𝑩 = 𝑩} lB {f ∘ lower}{lift} lABh λ _ _ → refl
+
+
+-- Monomorphisms and epimorphisms
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+         {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ)
+         where
+
+ private
+  A = 𝕌[ 𝑨 ]  -- carrier of domain of 𝑨
+  B = 𝕌[ 𝑩 ]
+
+ is-monomorphism : (A → B) → Type _
+ is-monomorphism g = is-homomorphism 𝑨 𝑩 g × IsInjective g
+
+ mon : Type(𝓞 ⊔ 𝓥 ⊔ α ⊔ β)
+ mon = Σ[ g ∈ (A → B) ] is-monomorphism g
+
+ is-epimorphism : (A → B) → Type _
+ is-epimorphism g = is-homomorphism 𝑨 𝑩 g × IsSurjective g
+
+ epi : Type _
+ epi = Σ[ g ∈ (A → B) ] is-epimorphism g
+
+
+-- The "hom reduct" of a mon
+-- N.B. 𝑨 explicit, 𝑩 implicit
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+         {β ρᵇ : Level} {𝑩 : SetoidAlgebra β ρᵇ} where
+
+ mon-to-hom : mon 𝑨 𝑩 → hom 𝑨 𝑩
+ mon-to-hom ϕ = ∣ ϕ ∣ , fst ∥ ϕ ∥
+
+
+-- The "hom reduct" of an epi
+-- N.B. 𝑨 implicit, 𝑩 explicit
+module _ {α ρᵃ : Level} {𝑨 : SetoidAlgebra α ρᵃ}
+         {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ) where
+
+ epi-to-hom : epi 𝑨 𝑩 → hom 𝑨 𝑩
+ epi-to-hom ϕ = ∣ ϕ ∣ , fst ∥ ϕ ∥
+
+\end{code}
+
+
+
+#### Kernels of homomorphisms for SetoidAlgebras
+
+\begin{code}
+
+
+module _ {α ρᵃ : Level} (𝑨 : SetoidAlgebra α ρᵃ)
+         {β ρᵇ : Level} (𝑩 : SetoidAlgebra β ρᵇ)
+         where
+ private
+  A = 𝕌[ 𝑨 ]
+  B = 𝕌[ 𝑩 ]
+
+
+ homker-comp : swelldef 𝓥 β → (h : hom 𝑨 𝑩) → 𝑨 ∣≈ (ker ∣ h ∣)
+ homker-comp wd h f {u}{v} kuv = ∣ h ∣((f ̂ 𝑨) u)   ≡⟨ ∥ h ∥ f u ⟩
+                                     (f ̂ 𝑩)(∣ h ∣ ∘ u) ≡⟨ wd(f ̂ 𝑩)(∣ h ∣ ∘ u)(∣ h ∣ ∘ v)kuv ⟩
```

Comment:
Something is distinctly odd here. I'm thinking something upstream doesn't take the Setoid approach seriously enough (i.e. is still too extensional) as extensionality shouldn't be needed.

Reply:
Okay, I'll try to resolve this now.

Reply:
Actually, this is not full function extensionality.  The kernel relation is (x , y) : h x = h y, and what we need here is to prove that if the tuples u and v are pointwise in the kernel of h (h (u i) = h (v i), for all i), then h ∘ u = h ∘ v.  However, we don't really need this equality to hold, we just need to know that f ̂ 𝑩 cannot distinguish between h ∘ u and  h ∘ v (that is, the pair (h ∘ u , h ∘ v) should be in the kernel of f ̂ 𝑩.  That's why we use the weaker version of funext, called swelldef.

I'm not sure whether/how Setoid features could be used to avoid this.